### PR TITLE
Fix coverage simplifier issues 953 962

### PR DESCRIPTION
**Pull request by mukoki**
Replace the previous pull request #961 : fix both issues #953 and #962 according to the discussion

### CODE REVIEW

**Comment by mukoki** [comment]
I tried to follow your suggestions about identification of nodes on shells sharing only one point.
Finally, this new pull request fix several bugs : 
- fix computation of CoverageEdge key in the case of ring including a single node (= first commit of my first PR)
- identifiying single node when two shells touch each other (propagate the information to CoverageEdge and Edge as a boolean) and pass it through MultiLineString via a BitSet included in the Geometry
- also fix PriorityQueue of corners which did not include all possible corners (I had to change the test testFilledHole() as the expected polygon was not simplified to the maximum before this change). Also clean the Queue from deprecated corners)
- add some tests and fix GeometryTestCase which was displaying the wrong message
- there is also an unfinished change in LinkedList (not sure it is necessary)

**Comment by mukoki** [comment]
On a significative dataset (117 Mb), the processing time goes from 41.5 s to 46 s after applying the PR.

File: modules/core/src/main/java/org/locationtech/jts/simplify/LinkedLine.java [hunk 1]
```diff
@@ -51,6 +51,7 @@ private int[] createNextLinks(int size) {
       next[i] = i + 1;
     }
     next[size - 1] = isRing ? 0 : size;
+    next[size - 1] = isRing ? 0 : NO_COORD_INDEX;
```
Comment (dr-jts) [inline modules/core/src/main/java/org/locationtech/jts/simplify/LinkedLine.java:54]: I think this is good

File: modules/core/src/main/java/org/locationtech/jts/coverage/TPVWSimplifier.java [hunk 2]
```diff
@@ -111,9 +108,11 @@ private List<Edge> createEdges(MultiLineString lines) {
     List<Edge> edges = new ArrayList<Edge>();
     if (lines == null)
       return edges;
+    BitSet freeRings = (lines.getUserData() != null && lines.getUserData() instanceof BitSet) ?
```
Comment (dr-jts) [inline modules/core/src/main/java/org/locationtech/jts/coverage/TPVWSimplifier.java:111]: Not crazy about passing this via userData.  Can't it be a parameter?

And I actually prefer your original idea of passing two sets of lines, one with endpoint constraints and one of the unconstrained rings.

File: modules/core/src/main/java/org/locationtech/jts/coverage/CoverageRingEdges.java [hunk 3]
```diff
@@ -216,6 +217,22 @@ private Set<Coordinate> findNodes(Geometry[] coverage) {
     return nodes;
   }
 
+  private Set<Coordinate> findBoundaryNodes(Set<LineSegment> lineSegments) {
+    Map<Coordinate, Integer> boundaryVertexCounter = new HashMap<>();
+    Set<Coordinate> nodes = new HashSet<Coordinate>();
+    for (LineSegment line : lineSegments) {
+      int count = boundaryVertexCounter.getOrDefault(line.p0, 0);
+      boundaryVertexCounter.put(line.p0, count+1);
+      count = boundaryVertexCounter.getOrDefault(line.p1, 0);
```
Comment (dr-jts) [inline modules/core/src/main/java/org/locationtech/jts/coverage/CoverageRingEdges.java:226]: Why set this twice?  Seems awkward.

Comment (dr-jts) [inline modules/core/src/main/java/org/locationtech/jts/coverage/CoverageRingEdges.java:226]: Definining a new local var would make this logic clearer

Comment (mukoki) [reply modules/core/src/main/java/org/locationtech/jts/coverage/CoverageRingEdges.java:226]: Do you mean a new local variable for the second count ? I can do it for sure. Not sure it would make the logic clearer though as the variable count all points with the same coordinate, either p0 or p1.
Indeed, I try to follow the JTS style which is mostly based on java5 I think ,but from my point of view, a more readable version would be : 
  private Set<Coordinate> findBoundaryNodes(Set<LineSegment> lineSegments) {
    Map<Coordinate, Integer> counter = new HashMap<>();
    for (LineSegment line : lineSegments) {
      counter.put(line.p0, counter.getOrDefault(line.p0, 0) + 1);
      counter.put(line.p1, counter.getOrDefault(line.p1, 0) + 1);
    }
    return counter.entrySet().stream()
        .filter(e->e.getValue()>2)
        .map(Map.Entry::getKey).collect(Collectors.toSet());
  }

Comment (dr-jts) [reply modules/core/src/main/java/org/locationtech/jts/coverage/CoverageRingEdges.java:226]: That looks even better!

**Comment by dr-jts** [comment]
You'll need to rebase, since I pushed a fix to `GeometryTestCase`.

File: modules/core/src/main/java/org/locationtech/jts/coverage/TPVWSimplifier.java (hunk 2 above)
Comment (mukoki) [reply modules/core/src/main/java/org/locationtech/jts/coverage/TPVWSimplifier.java:111]: Can probably do that, but this separation into two sets of lines is probably of no use for the end user. That's why I thought it would be better to consider it as an implementation detail. I understand the argument to keep things accessible and to avoid hidden tricks though.

Comment (dr-jts) [reply modules/core/src/main/java/org/locationtech/jts/coverage/TPVWSimplifier.java:111]: Ok, either way is fine for now.  If a different calling signature becomes useful it can always be added later.

**Comment by mukoki** [comment]
Sorry, with the different commits, sync and rebase, I'm not sure the PR is still usable for you. Let me know. I've also made a separate PR where I replaced the BitSet trick by a second MultiLineString to pass the unconstrained rings. Use the one you prefer or another way if you have a precise idea in mind.

File: modules/core/src/main/java/org/locationtech/jts/coverage/TPVWSimplifier.java (hunk 2 above)
Comment (dr-jts) [reply modules/core/src/main/java/org/locationtech/jts/coverage/TPVWSimplifier.java:114]: Actually I should have said - the BitSet data is fine, but would be better to pass it as an actual parameter to `TPVWSimplifier`.  I prefer not to hide key data in the untyped `userData` field.

**Comment by dr-jts** [comment]
This PR looks good, with one small change request as noted in a comment (pass the BItSet as an explicit parameter to `TPVWSImplifier`).
